### ISSUE TITLE
Dollar sign in environment variable crashes plugin

### ISSUE THREAD

**Issue by mb720** [Tapad/sbt-docker-compose#83]
Hi!

The following line in docker.compose.yml trips up the plugin:
```
    environment:
      - TLS_KEY_STORE_PASSWORD=${TLS_KEY_STORE_PASSWORD:-9f0ht032fr09fds909SDG$3gt32f#@FDSfs}
```

Console output with stack trace:

> Reading Compose File: C:\Users\maine\Dev\Scala\scala-js-example\app\jvm/../../docker-compose.yml
> java.lang.IndexOutOfBoundsException: No group 3
>         at java.util.regex.Matcher.start(Matcher.java:375)
>         at java.util.regex.Matcher.appendReplacement(Matcher.java:880)
>         at java.util.regex.Matcher.replaceAll(Matcher.java:955)
>         at java.lang.String.replaceAll(String.java:2223)
>         at com.tapad.docker.ComposeFile$$anonfun$processVariableSubstitution$1.apply(ComposeFile.scala:340)
>         at com.tapad.docker.ComposeFile$$anonfun$processVariableSubstitution$1.apply(ComposeFile.scala:339)
>         at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:144)
>         at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:144)
>         at scala.collection.Iterator$class.foreach(Iterator.scala:727)
>         at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
>         at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:144)
>         at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1157)
>         at com.tapad.docker.ComposeFile$class.processVariableSubstitution(ComposeFile.scala:339)
>         at com.tapad.docker.DockerComposePluginLocal.processVariableSubstitution(DockerComposePlugin.scala:89)
>         at com.tapad.docker.ComposeFile$class.readComposeFile(ComposeFile.scala:311)
>         at com.tapad.docker.DockerComposePluginLocal.readComposeFile(DockerComposePlugin.scala:89)
>         at com.tapad.docker.DockerComposePluginLocal.startDockerCompose(DockerComposePlugin.scala:248)
>         at com.tapad.docker.DockerComposePluginLocal.launchInstanceWithLatestChanges(DockerComposePlugin.scala:181)
>         at com.tapad.docker.DockerComposePluginLocal.restartRunningInstance(DockerComposePlugin.scala:204)
>         at com.tapad.docker.DockerComposePluginLocal$$anonfun$dockerComposeRestartCommand$2.apply(DockerComposePlugin.scala:130)
>         at com.tapad.docker.DockerComposePluginLocal$$anonfun$dockerComposeRestartCommand$2.apply(DockerComposePlugin.scala:128)
>         at sbt.Command$$anonfun$applyEffect$1$$anonfun$apply$2.apply(Command.scala:59)
>         at sbt.Command$$anonfun$applyEffect$1$$anonfun$apply$2.apply(Command.scala:59)
>         at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:61)
>         at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:61)
>         at sbt.Command$.process(Command.scala:93)
>         at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:96)
>         at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:96)
>         at sbt.State$$anon$1.process(State.scala:184)
>         at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:96)
>         at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:96)
>         at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
>         at sbt.MainLoop$.next(MainLoop.scala:96)
>         at sbt.MainLoop$.run(MainLoop.scala:89)
>         at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:68)
>         at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:63)
>         at sbt.Using.apply(Using.scala:24)
>         at sbt.MainLoop$.runWithNewLog(MainLoop.scala:63)
>         at sbt.MainLoop$.runAndClearLast(MainLoop.scala:46)
>         at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:30)
>         at sbt.MainLoop$.runLogged(MainLoop.scala:22)
>         at sbt.StandardMain$.runManaged(Main.scala:57)
>         at sbt.xMain.run(Main.scala:29)
>         at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:109)
>         at xsbt.boot.Launch$.withContextLoader(Launch.scala:128)
>         at xsbt.boot.Launch$.run(Launch.scala:109)
>         at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:35)
>         at xsbt.boot.Launch$.launch(Launch.scala:117)
>         at xsbt.boot.Launch$.apply(Launch.scala:18)
>         at xsbt.boot.Boot$.runImpl(Boot.scala:41)
>         at xsbt.boot.Boot$.main(Boot.scala:17)
>         at xsbt.boot.Boot.main(Boot.scala)
> [error] java.lang.IndexOutOfBoundsException: No group 3

With trial and error I isolated the character that causes the exception: It's the dollar sign `$` in the password.

Yet when I tried to simplify the line to 
`      - A_SECRET=${ENV_VAR_WITH_SECRET:-secret$password}`

I merely got

> Illegal group reference

in the console, no stack trace.

Thanks in advance!

**Comment by kurtkopchik** [comment]
Hi @mb720 - I made a change https://github.com/Tapad/sbt-docker-compose/commit/85a2136c8c99e0332f77d41ce183c8fa87bd1393 to better handle '$' characters when they occur in the variable substitution value.

I published the change to `1.0.24-SNAPSHOT`.  Let me know if it works for you and I can get it into the next official release.

```
resolvers += "Sonatype Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots/"
addSbtPlugin("com.tapad" % "sbt-docker-compose" % "1.0.24-SNAPSHOT")
```

Thanks again for the feedback!
